File: leetcode_ly/c++/UnionFind/NumberOfIslands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int row = grid.size();
        if(row == 0) return 0;
        int col = grid[0].size();
        int ans = 0;
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if(grid[i][j] == '1'){
                    dfs(grid,i,j);
                    ans++;
                }
            }
        }
        return ans;
    }
    
    void dfs(vector<vector<char>>& grid,int i,int j){
        int m = grid.size();
        int n = grid[0].size();
        if(i < 0 || j < 0 || i >= m || j >= n){
            return;
        }
        
        if(grid[i][j] == '1'){
            grid[i][j] = '2';
            dfs(grid,i - 1,j);
            dfs(grid,i + 1,j);
            dfs(grid,i,j - 1);
            dfs(grid,i,j + 1);
        }
    }
};

File: leetcode_ly/java/Depth-first && Breadth-first/NumberOfIslands.java
public class Solution {
    static int m;
    static int n;
    public int numIslands(char[][] grid) {
        if(grid == null || grid.length == 0) return 0;
        m = grid.length;
        n = grid[0].length;
        
        int count = 0;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == '1'){
                    dfs(grid,i,j);
                    count++;
                }
            }
        }
        return count;
    }
    
    public void dfs(char[][] grid,int row,int col){
        if(row < 0 || row >= m || col < 0 || col >= n || grid[row][col] != '1') return;
        //将同一个岛中的1全部变为0
        grid[row][col] = '0';
        
        dfs(grid,row - 1,col);
        dfs(grid,row + 1,col);
        dfs(grid,row,col - 1);
        dfs(grid,row,col + 1);
    }
}